### ISSUE TITLE
Does calling get_redis_connection() multiple times open multiple connections?

### ISSUE THREAD

**Issue by alexington** [niwinz/django-redis#204]
My django application makes multiple calls to `get_redis_connection()` in the course of each web request.  I would say it's around 3-6 calls to `get_redis_connection()` per request.

Should I be concerned about any potential resource issues?  Are the connections to the redis client cached within a single django web request?  Or does every call to `get_redis_connection()` create a new connection to the redis client?

Sorry if this looks like a support question!  I just don't know how to test this on my own, or where else to ask.  I'd love to have a better understanding of what's going on each time I use this to get raw client access.  Thanks a lot!

**Comment by niwinz** [comment]
Each time that you call to get_redis_connection function new `StrictRedis` instance is created; but that does not implies that new open connection to redis is open.

This is a mostly exact step list of that happens when you call that function:

- New redis-py compatible parameters map is created from the django configuration for the specified alias.
- New connection pool is created for that parameters (only if it already does not exists, this step is skiped if the pool is already exists for that parameters).
- New StrictRedis instance is created with the connection pool (from step 2).

Until now no new connection is open and no connection are aquired from the connection pool.

The connection is acquired from the pool (and created if pool was empty) only when you perform an operation and immediately returned when your operation is finished (with operation, I mean any command execution). There are spesical commands such as pub sub that acquires connection for the whole operation and you should take care that performing pubsub maintains a persistent connection so that connection is not returned to the pool until you finish the pubsub operation.

So, in summary, if you call `get_redis_connection` many times, you will produce a small amount of "garbage" instances of `StrictRedis` class but in the best of the cases you will only create only one connection (if no previously created one is avaliable in the pool) in the request lifecycle. The worst case it depends on other stuff where you are using redis on your application...

**Comment by alexington** [comment]
Thank you, that was extremely helpful!

I was previously considering passing and reusing the same `StrictRedis` instance throughout the application.  But I think you're saying that wouldn't do much at all. At least it wouldn't affect the number of connections since a single `StrictRedis` instance is not tied to a specific connection.  Instead, connections are opened (or acquired from the pool) when operations (get, set, etc.) are executed.

I just had one last question.  Is connection pooling occurring out of the box?

From the [redis-py docs](https://github.com/andymccurdy/redis-py#connection-pools):
> Behind the scenes, redis-py uses a connection pool to manage connections to a Redis server. By default, each Redis instance you create will in turn create its own connection pool.

From the [django-redis docs](http://niwinz.github.io/django-redis/latest/#_connection_pools):
> Behind the scenes, django-redis uses the underlying redis-py connection pool implementation... The default redis-py behavior is to not close connections, recycling them when possible. (django-redis docs)

This appears to say that no additional setup is required to take advantage of connection pooling.  Is that correct?

**Comment by niwinz** [comment]
Yes, this is correct, connection pooling is the default approach and can only be overwritten creating own connection factory and specify it to django-redis.

**Comment by alexington** [comment]
Awesome, thanks again for the replies.  Appreciate it.
